pkg/cloudprovider/provider/hetzner/provider.go

```diff
@@ -176,6 +182,34 @@ func (p *provider) Create(machine *v1alpha1.Machine, userdata string) (instance.
 		return nil, hzErrorToTerminalError(err, "failed to get server type")
 	}
 
+	tmpRSAKeyPair, err := rsa.GenerateKey(rand.Reader, privateRSAKeyBitSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create private RSA key: %v", err)
+	}
+
+	if err := tmpRSAKeyPair.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate private RSA key: %v", err)
+	}
+	pubKey, err := ssh.NewPublicKey(&tmpRSAKeyPair.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate ssh public key: %v", err)
+	}
```

Comment:
We have the same logic now twice, fir Hetzner and Digitalocean.
Better extract it into a separate package, which returns a ssh.PublicKey & a random name.